compiler/parser/RdrHsSyn.hs

```diff
@@ -3084,6 +3085,80 @@ splitHsAppTysPs t = go t []
     go (L _ (HsAppTy _ f a)) as = go f (a : as)
     go f                     as = (f, as)
 
+------------
+
+-- | Desugar a try–catch expressions for exception-handling.
+
+mkTryCatchExpr
+  :: LHsExpr GhcPs
+  -> Located ([AddAnn],[LMatch GhcPs (LHsExpr GhcPs)])
+  -> P (HsExpr GhcPs)
+mkTryCatchExpr tryExpr@(L tryLoc _) (L catchLoc (_, rawAlts)) = do
+
+  let tryCatchVar = mkQualVar $ mkVarOcc "_tryCatch"
+      fromAnyExceptionVar = mkQualVar $ mkVarOcc "fromAnyException"
+
+      wildCard :: LPat GhcPs
+      wildCard = noLoc (WildPat noExt)
+
+      emptyBinds :: LHsLocalBinds GhcPs
+      emptyBinds = noLoc (EmptyLocalBinds noExt)
+
+      tryMatch :: LMatch GhcPs (LHsExpr GhcPs)
+      tryMatch = noLoc
+          $ Match noExt LambdaExpr [wildCard] Nothing
+          $ GRHSs noExt [noLoc $ GRHS noExt [] tryExpr] emptyBinds
+      tryLambda = L tryLoc $ HsLam noExt $ MG noExt (noLoc [tryMatch]) Generated
+
+      -- We need to convert the patterns in catch expression from
+      --    "PATTERN -> EXPR"
+      -- into
+      --    "(fromAnyException -> Some PATTERN) -> Some EXPR".
+      -- The following bindings achieve this.
+
+      report :: SrcSpan -> String -> P a
+      report loc e = addFatalError loc (text e)
+
+      convertLPat :: LPat GhcPs -> P (LPat GhcPs)
+      -- TODO https://github.com/digital-asset/daml/issues/8020
+      --   Emit helpful error for single variable or wildcard pattern
+      --   without type annotation.
+      convertLPat p = pure . noLoc . ViewPat noExt fromAnyExceptionVar . noLoc
+          $ ConPatIn (noLoc . qualifyDesugar $ mkDataOcc "Some") (PrefixCon [p])
+
+      convertLGRHS :: LGRHS GhcPs (LHsExpr GhcPs) -> P (LGRHS GhcPs (LHsExpr GhcPs))
+      convertLGRHS (L loc (GRHS ext guards body)) = do
+          let body' = mkSome body
+          pure (L loc (GRHS ext guards body'))
+      convertLGRHS (L loc _) =
+          report loc "Unexpected right-hand side for catch."
+
+      convertGRHSs :: SrcSpan -> GRHSs GhcPs (LHsExpr GhcPs) -> P (GRHSs GhcPs (LHsExpr GhcPs))
+      convertGRHSs _ (GRHSs ext lgrhss localBinds) = do
+          lgrhss' <- mapM convertLGRHS lgrhss
+          pure (GRHSs ext lgrhss' localBinds)
+      convertGRHSs loc _ = do
+          report loc "Unexpected right-hand side for catch."
+
+      convertMatch :: LMatch GhcPs (LHsExpr GhcPs) -> P (LMatch GhcPs (LHsExpr GhcPs))
+      convertMatch (L loc (Match ext _ pats _ grhss)) = do
+          pats' <- mapM convertLPat pats
+          grhss' <- convertGRHSs loc grhss
+          pure (L loc (Match ext CaseAlt pats' Nothing grhss'))
```

Comment:
Maybe add some comments on what these do? It’s a bit hard to see spot just from looking through the code.